File: backend/app/schemas/conversation.py
"""Pydantic schemas for conversation API."""

from typing import Optional, List

from pydantic import BaseModel, Field


class GrammarCorrection(BaseModel):
    """Grammar correction details."""

    original: str = Field(..., description="The original incorrect phrase")
    corrected: str = Field(..., description="The corrected phrase")
    explanation: str = Field(..., description="Explanation of the correction")


class ConversationStartRequest(BaseModel):
    """Request to start a new conversation."""

    topic: Optional[str] = Field(None, description="Conversation topic or theme")
    target_words: Optional[list[str]] = Field(
        None, description="Vocabulary words to practice during conversation"
    )
    scenario: Optional[str] = Field(
        None, description="Scenario ID for role-play mode (e.g., 'job_interview', 'restaurant')"
    )


class ConversationStartResponse(BaseModel):
    """Response when starting a new conversation."""

    session_id: str = Field(..., description="Unique session identifier")
    opening_message: str = Field(..., description="Opening message from the assistant")
    topic: Optional[str] = Field(None, description="The conversation topic")
    target_words: Optional[list[str]] = Field(None, description="Words to practice")
    scenario: Optional[str] = Field(None, description="Active scenario ID if in role-play mode")
    scenario_name: Optional[str] = Field(None, description="Human-readable scenario name")
    user_role: Optional[str] = Field(None, description="User's role in the scenario")
    user_goal: Optional[str] = Field(None, description="User's goal in the scenario")


class ConversationMessageRequest(BaseModel):
    """Request to send a message in a conversation."""

    message: str = Field(..., min_length=1, description="The user's message")
    session_id: Optional[str] = Field(None, description="Session ID for continuing conversation")


class ConversationResponse(BaseModel):
    """Response from the conversation agent."""

    reply: str = Field(..., description="The assistant's reply")
    followUp: str = Field("", description="A follow-up question to continue the conversation")
    correction: Optional[GrammarCorrection] = Field(
        None, description="Grammar correction if the user made a mistake"
    )
    session_id: Optional[str] = Field(None, description="Session ID for the conversation")


class ConversationMessage(BaseModel):
    """A single message in a conversation."""

    id: str = Field(..., description="Unique message identifier")
    role: str = Field(..., description="Message role: 'user' or 'assistant'")
    content: str = Field(..., description="Message content")
    correction: Optional[GrammarCorrection] = Field(
        None, description="Grammar correction if applicable"
    )


class ConversationFeedbackRequest(BaseModel):
    """Request for conversation feedback."""

    session_id: str = Field(..., description="Session ID to get feedback for")


class ConversationFeedbackResponse(BaseModel):
    """Response with conversation feedback."""

    feedback: str = Field(..., description="Feedback summary for the conversation")
    session_id: str = Field(..., description="Session ID")


# Scenario schemas
class ScenarioInfo(BaseModel):
    """Information about a conversation scenario."""

    id: str = Field(..., description="Unique scenario identifier")
    name: str = Field(..., description="Human-readable scenario name")
    role: str = Field(..., description="AI's role in the scenario")
    userRole: str = Field(..., description="User's role in the scenario")
    userGoal: str = Field(..., description="User's goal to achieve")
    vocabulary: List[str] = Field(..., description="Scenario-specific vocabulary words")


class ScenariosListResponse(BaseModel):
    """Response with list of available scenarios."""

    scenarios: List[ScenarioInfo]

File: src/lib/api/conversation.ts
/**
 * Conversation API functions
 */

import { api } from '../api';

// Types matching backend schemas
export interface GrammarCorrection {
  original: string;
  corrected: string;
  explanation: string;
}

export interface ConversationMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  correction?: GrammarCorrection;
}

export interface ScenarioInfo {
  id: string;
  name: string;
  role: string;
  userRole: string;
  userGoal: string;
  vocabulary: string[];
}

export interface ConversationStartResponse {
  session_id: string;
  opening_message: string;
  topic?: string;
  target_words?: string[];
  scenario?: string;
  scenario_name?: string;
  user_role?: string;
  user_goal?: string;
}

export interface ConversationOptions {
  topic?: string;
  targetWords?: string[];
  scenario?: string;
}

export interface ConversationResponse {
  reply: string;
  followUp: string;
  correction?: GrammarCorrection;
  session_id?: string;
}

export interface ConversationFeedbackResponse {
  feedback: string;
  session_id: string;
}

// Session storage for conversation
let currentSessionId: string | null = null;

/**
 * Get available conversation scenarios
 */
export async function getScenarios(): Promise<ScenarioInfo[]> {
  const response = await api.get<{ scenarios: ScenarioInfo[] }>('/api/conversation/scenarios');
  return response.scenarios;
}

/**
 * Start a new conversation session
 */
export async function startConversation(
  options?: ConversationOptions
): Promise<ConversationStartResponse> {
  const response = await api.post<ConversationStartResponse>('/api/conversation/start', {
    topic: options?.topic,
    target_words: options?.targetWords,
    scenario: options?.scenario,
  });
  currentSessionId = response.session_id;
  return response;
}

/**
 * Send a message in the conversation
 */
export async function sendMessage(
  message: string,
  sessionId?: string
): Promise<ConversationResponse> {
  const response = await api.post<ConversationResponse>('/api/conversation/message', {
    message,
    session_id: sessionId || currentSessionId,
  });

  if (response.session_id) {
    currentSessionId = response.session_id;
  }

  return response;
}

/**
 * Get the opening message for a new conversation
 * This starts a new session and returns the opening message
 */
export function getOpeningMessage(): ConversationMessage {
  // Start a new session in the background
  startConversation().then(response => {
    currentSessionId = response.session_id;
  }).catch(console.error);

  // Return a default opening message while the session starts
  const openings = [
    "Hi there! I'm excited to practice English with you today. So, what did you do this morning?",
    "Hey! Great to meet you. Let's have a nice chat. What's your favorite way to spend a free afternoon?",
    "Hello! I'm here to help you practice English. Tell me, what's something you're looking forward to this week?",
  ];

  return {
    id: 'opening',
    role: 'assistant',
    content: openings[Math.floor(Math.random() * openings.length)],
  };
}

/**
 * Generate feedback for the conversation
 */
export async function generateFeedback(sessionId?: string): Promise<string> {
  const response = await api.post<ConversationFeedbackResponse>('/api/conversation/feedback', {
    session_id: sessionId || currentSessionId,
  });
  return response.feedback;
}

/**
 * Get current session ID
 */
export function getCurrentSessionId(): string | null {
  return currentSessionId;
}

/**
 * Clear current session
 */
export function clearSession(): void {
  currentSessionId = null;
}

export default {
  getScenarios,
  startConversation,
  sendMessage,
  getOpeningMessage,
  generateFeedback,
  getCurrentSessionId,
  clearSession,
};
